### ISSUE TITLE
FR. DoT keepalive

### ISSUE THREAD

**Issue by charlieporth1** [folbricht/routedns#220]
FR. DoT TCP keepalive if it's not there already

**Comment by cbuijs** [comment]
I think this would/could apply on `tcp` (unencrypted) dns as well.

**Comment by charlieporth1** [comment]
> I think this would/could apply on `tcp` (unencrypted) dns as well.

Good point!

**Comment by charlieporth1** [comment]
Ontop of that does routedns reuse TCP connections?

**Comment by folbricht** [comment]
It does reuse connections. Keepalive would have to be configured with quite a high frequency, many servers close connections after 2s of inactivity

**Comment by charlieporth1** [comment]
@folbricht we could put that in the README or check if the server has it enabled

**Comment by charlieporth1** [comment]
@folbricht is there a reason we are not using `ReusePort: true,` in the DNS listener provided by `github.com/miekg/dns`? 
It could help according to this [link](https://www.google.com/search?q=is+resuing+ports+faster&newwindow=1&rlz=1C1GCEU_enUS978US978&sxsrf=APq-WBtYZJgD-BmfjmjJJ38ZGPvWynt8rw%3A1647442810868&ei=evsxYrzNNOHL0PEPme-Q2A4&ved=0ahUKEwi88p3c8sr2AhXhJTQIHZk3BOsQ4dUDCA8&uact=5&oq=is+resuing+ports+faster&gs_lcp=Cgdnd3Mtd2l6EAMyBwghEAoQoAEyBwghEAoQoAE6BwgjEOoCECc6BAgjECc6BQgAEJECOg4ILhCABBCxAxDHARDRAzoOCC4QgAQQsQMQxwEQowI6CAgAELEDEIMBOggIABCABBCxAzoLCAAQgAQQsQMQgwE6BAgAEEM6CwguELEDEIMBENQCOgsILhCABBCxAxDUAjoICC4QgAQQsQM6BQgAEIAEOgkIABAKEEYQ-QE6BAgAEAo6BAgAEA06BggAEA0QHjoICAAQCBANEB46BQgAEIYDOggIIRAWEB0QHkoECEEYAEoECEYYAFAAWOA1YM42aAVwAXgAgAGDAYgBkxSSAQQ5LjE1mAEAoAEBsAEKwAEB&sclient=gws-wiz)

**Comment by folbricht** [comment]
Port reuse would only be necessary if routedns didn't reuse connections and opened TCP connections for every query, so I don't think it's going to improve performance at all at this point.

**Comment by charlieporth1** [comment]
Now I could be wrong but is routedns reusing ports for non routedns connections (server side and non routedns clients)? It looks like its only on the client-side of things and not the server-side it would also seem to be the cause of this issue according to [this](https://hea-www.harvard.edu/~fine/Tech/addrinuse.html) however that issue is UDP so I don't know

**Comment by charlieporth1** [comment]
@folbricht the reason I say specifically routedns only being client-side initiating port reuse is that many of my devices that use DoT are Android which doesn't run routedns natively.

**Comment by charlieporth1** [comment]
@folbricht Just checking in

**Comment by folbricht** [comment]
The `ReusePort` option in https://pkg.go.dev/github.com/miekg/dns#Server doesn't actually impact the number of sockets etc. It allows running multiple servers on the same port, so "reuse" of the same port for more than one server listener. That really shouldn't change behavior at all here. What might be more useful here would be tuning socket-flags in the OS that routedns runs on, like make it free up sockets faster.

**Comment by charlieporth1** [comment]
@folbricht Interesting. What do you mean when you say `multiple servers` - like physical machines or software programs or what?

**Comment by folbricht** [comment]
Yes, so that flag allows running more than one server processs/listener on the same port. So normally if you bring up a server on port 53 and then tried to bring up another one it'd fail saying "port already in use" The flag allows multiple listeners using the same port number basically. I don't think this is actually needed.

Haven't had time to debug this, but what could be happening is that when connections terminate the sockets are kept for a while longer by the OS. Busy servers that deal with a lot of connection open/close are typically optimized for this, for example clearing up those old sockets much faster, making room for new ones. That would be on the OS-level though. It's also possible there's a bug in routedns where it doesn't properly close connections. I'd need to be able to reproduce the issue to debug as I haven't seen it myself. Either my config is different or my server isn't busy enough (likely).

**Comment by charlieporth1** [comment]
@folbricht I was able to create a script that generates traffic and will cause this issue if you want it.

FYI it does pretty much spam the server with DNS requests kinda like a DDoS so be careful with it and I am not going to post it publicly

**Comment by charlieporth1** [comment]
Also as far as TCP keep alive does [IdleTimeout](https://pkg.go.dev/github.com/miekg/dns#Server) in does a similar if not the same thing as TCP keepalive

**Comment by folbricht** [comment]
Can you include the routedns config you use to repro the issue? Also, do you open a new new client connection for every query when running the script or is the client reusing a connection?

**Comment by charlieporth1** [comment]
It's pretty much a bash script that uses dig
